### ISSUE TITLE
False positives when defering import of `typing.Literal`

### ISSUE THREAD

**Issue by mikeshardmind** [astral-sh/ruff#15306]
```py
# _typings.py

from __future__ import annotations

TYPE_CHECKING = False

if TYPE_CHECKING:
    from typing import Any, Final, Literal, Self, final
else:
    def final(f):  # noqa: ANN001
        return f


def __getattr__(name: str):
    if name == "final":  # this one actually executes at runtime
        return final

    if name in {"Any", "Final", "Literal", "Self"}:
        import typing
        return getattr(typing, name)

    msg = f"module {__name__!r} has no attribute {name!r}"
    raise AttributeError(msg)


__all__ = ["Any", "Final", "Literal", "Self", "final"]
```

```py
# any other module

from ._typings import Literal

type HandleableSignals = Literal["SIGINT", "SIGTERM", "SIGBREAK", "SIGHUP"]
```

This causes ruff to give an F821 (Undefined name) on each string literal here, while also trying to have me remove the quotes with TC0008 This happens with that as the entirety of the file, and there is no corresponding symbol with that name here. 

(This is used to cross-platform handle any of these signals so long as they are handleable on the platform, as they may not exist on signals.Signals at runtime on a platform, this pulls the intersections of what's on the system, and which of the ones the user wants handled)

**Comment by AlexWaygood** [comment]
Thanks for the issue -- have you tried making use of this setting? https://docs.astral.sh/ruff/settings/#lint_typing-modules

**Comment by mikeshardmind** [comment]
Setting handles it, and this is documented behavior, so I'm going to close this, but this is pretty surprising to need to specify.

**Comment by AlexWaygood** [comment]
Yeah, I agree that this is often somewhat confusing for users; sorry about that. In the long term we hope to transform Ruff into a linter capable of multifile analysis, which will solve this problem, but it'll be a long road to get there unfortunately.
